blogposts/2023-02-10-adding-snapshot-testing-to-rnw-apps.md

```diff
@@ -0,0 +1,93 @@
+
+To begin, I followed Jest's [Getting Started](https://jestjs.io/docs/tutorial-react-native) guide for adding testing to React Native apps. This gave me a good overview of potential Jest configurations/customizations, how to write a simple snapshot test, and information on snapshot reference files.
+
+Writing snapshot tests in Jest is fairly simple. Once you have your `App-test.js` file, you'll want to import the desired component that you want to snapshot and write a test case. Here is an example snippet from React Native Gallery's testing of its Button component page:
+```js
+import {ButtonExamplePage} from '../src/examples/ButtonExamplePage';
+
+test('Button Example Page', () => {
+  const tree = create(<ButtonExamplePage />).toJSON();
+  expect(tree).toMatchSnapshot();
+});
+```
+
+As I began to develop, I realized there was some additional configuration I needed to add in order to get my tests to run correctly. First, I had several modules that needed to be added to the `transformIgnorePatterns` Jest customization. This customization is used to specify which files should be transformed by Babel. By default, Jest only processes the app's and react-native's source code. You'll know if a module needs to be transformed, if the app throws an error saying "Jest encountered an unexpected token". In this case, the module should be listed in the `transformIgnorePatterns` customization within your app's `package.json`. See Jest's [transformIgnorePatterns](https://jestjs.io/docs/tutorial-react-native#transformignorepatterns-customization) documentation for syntax info.
+
+Next, I had several modules which required [mocking](https://jestjs.io/docs/mock-function-api/). Mock functions are often needed for modules which implement API's instead of UI components. Examples within React Native Gallery include the `react-native-device-info` module or the `react-navigation` module. If a function needs to be mocked by Jest, you'll commonly receive an error that the method is undefined, when trying to render a component which makes use of the method. Here's an example error for a `react-native-device-info` method that needs mocking: "`@react-native-community/react-native-device-info`: `NativeModule.RNDeviceInfo` is null". To add mocks to your Jest configuration, you'll want to create a `jest-setup.js` file at the root of your repo. Then you'll want to edit your Jest configuration in your app's `package.json` to identify that you've created a setup file. See the `package.json` source snippet below for syntax. Now you'll want to add contents to your `jest-setup.js` file. Some community modules already have `mock.js` files implemented in their source. If this is the case, you'll just want to reference this file in your setup file.
```

Comment:
```suggestion
Next, I had several modules which required [mocking](https://jestjs.io/docs/mock-function-api/). Mock functions are often needed for modules which implement API's instead of UI components. Examples within React Native Gallery include the `react-native-device-info` module or the `react-navigation` module. If a function needs to be mocked by Jest, you'll commonly receive an error that the method is undefined, when trying to render a component which makes use of the method. Here's an example error for a `react-native-device-info` method that needs mocking: "`@react-native-community/react-native-device-info`: `NativeModule.RNDeviceInfo` is null". To add mocks to your Jest configuration, you'll want to create a `jest-setup.js` file at the root of your repo. Then you'll want to edit your Jest configuration in your app's `package.json` to identify that you've created a setup file (*see the `package.json` source snippet below for syntax*). Now you'll want to add contents to your `jest-setup.js` file. Some community modules already have `mock.js` files implemented in their source. If this is the case, you'll just want to reference this file in your setup file.
```